### PR TITLE
🎉 Destination Meilisearch: migrating to python and fix integration 

### PR DESCRIPTION
**Pull request by cirdes**
## What
The connector was not working with last meilisearch version due a change in authorization method https://github.com/airbytehq/airbyte/issues/13076
Update meilisearch logo

## How
Meilisearch Java sdk was outdated so in order to fix the problem I had to migrate to python

## Recommended reading order
1. `spec.yaml`
2. `destionation.py`
3. `writer.py`

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

```
==================================================================================================== test session starts =====================================================================================================
platform darwin -- Python 3.9.11, pytest-6.1.2, py-1.11.0, pluggy-0.13.1 -- /Users/cirdes/Workspace/airbyte/airbyte-integrations/connectors/destination-meilisearch/.venv/bin/python
cachedir: .pytest_cache
rootdir: /Users/cirdes/Workspace/airbyte, configfile: pytest.ini
collected 2 items

unit_tests/unit_test.py::test_queue_write_operation PASSED
unit_tests/unit_test.py::test_flush PASSED
```

</details>

<details><summary><strong>Integration</strong></summary>

❯ python -m pytest -s -vv integration_tests/
```
==================================================================================================== test session starts =====================================================================================================
platform darwin -- Python 3.9.11, pytest-6.1.2, py-1.11.0, pluggy-0.13.1 -- /Users/cirdes/Workspace/airbyte/airbyte-integrations/connectors/destination-meilisearch/.venv/bin/python
cachedir: .pytest_cache
rootdir: /Users/cirdes/Workspace/airbyte, configfile: pytest.ini
collected 3 items

integration_tests/integration_test.py::test_check_valid_config PASSED
integration_tests/integration_test.py::test_check_invalid_config {"type": "LOG", "log": {"level": "ERROR", "message": "Check connection failed. Error: MeiliSearchApiError. 404 Client Error: Not Found for url: https://www.meilisearch.com/indexes"}}
PASSED
integration_tests/integration_test.py::test_write PASSED

===================================================================================================== 3 passed in 10.82s =====================================================================================================
```

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

Screen Shot 2022-10-22 at 3 05 58 PM

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Thanks for the contribution @cirdes someone will review this during the week!

**Comment by marcosmarxm** [comment]
Hello @cirdes I asked the connector team about your contribution if moving MeiliSearch from Java is beneficial or not. ASAP I get an answer will return to you.

**Comment by cirdes** [comment]
@marcosmarxm, I had to migrate from Java to Python because the lastest version of java sdk is not working and I was trying to use Meilisearch with airbyte. You can check [here](https://github.com/airbytehq/airbyte/issues/13076)

The last update on Java SDK was on [May 09](https://github.com/meilisearch/meilisearch-java/releases) and [Python SDK](https://github.com/meilisearch/meilisearch-python/releases) is been updated every week.

I don't see the point of keep in Java if the Meilisearch Java sdk is broken

**Comment by marcosmarxm** [comment]
> @marcosmarxm, I had to migrate from Java to Python because the lastest version of java sdk is not working and I was trying to use Meilisearch with airbyte. You can check [here](https://github.com/airbytehq/airbyte/issues/13076)
> 
> The last update on Java SDK was on [May 09](https://github.com/meilisearch/meilisearch-java/releases) and [Python SDK](https://github.com/meilisearch/meilisearch-python/releases) is been updated every week.
> 
> I don't see the point of keep in Java if the Meilisearch Java sdk is broken

This is the same opinion of connector team. I'll review your contribution as soon as possible :)

File: airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            time.sleep(timer)
+            if timer < 60:
+                timer = timer * 2
+            try:
+                task = self.client.get_task(task_uid)
+                status = task["status"]
+                if status == "succeeded" or status == "failed":
+                    break
+            except:
+                timer = 60
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py:47]:
```suggestion
    def wait_for_job(self, task_uid: str):
        timer = 1
        while timeout < timeout_limit:
            time.sleep(timer)
            timer = min(timer * 2, 60)
            task = self.client.get_task(task_uid)
            status = task["status"]
            if status == MeiliSearchTaskStatus.SUCCEDED or status == MeiliSearchTaskStatus.FAILED:
                break
```
Please simplify the function, if you're going to do a try/except you must raise the error you're dealing with.
Maybe take a look in the backoff library to apply the timer.
I think if the task == 'failed' you can log the reason why it failed.
Would be nice to have a timeout_limit to not have an infinity loop here.
Add unit tests to all functions for the writer.

File: airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        if len(self.write_buffer) == self.flush_interval:
+            self.flush()
+
+    def flush(self):
+        if len(self.write_buffer) == 0:
+            return
+
+        response = self.client.index(self.steam_name).add_documents(self.write_buffer)
+        self.wait_for_job(response.task_uid)
+        self.write_buffer.clear()
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py:33]: Please add a logger message saying you're flushing the records.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py (hunk 1 above)
Comment (cirdes) [reply airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py]: @marcosmarxm meilisearch sdk has a method to wait for the async operation to finished. Now is much simpler.
I added the unit tests and logger massage as well. 

The [destination docs](https://docs.airbyte.com/connector-development/tutorials/building-a-python-destination) refers the KvDB destination connector but it lacks unit tests and logger message on flush method.

File: airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py (hunk 2 above)
Comment (cirdes) [reply airbyte-integrations/connectors/destination-meilisearch/destination_meilisearch/writer.py]: Added!

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
<s>@cirdes I'm having problems setting up a local instance of MeiliSearch. Can you add a bootstrap document explaining how to setup a local instance and generate the API Key.

The error I got:
`Check connection failed. Error: No connection adapters were found for 'localhost:7700/indexes'`
</s>
My bad using the wrong host.

**Comment by cirdes** [comment]
@marcosmarxm , do you need anything else?

**Comment by marcosmarxm** [comment]
> @marcosmarxm , do you need anything else?

I need to setup a MeiliSearch instance to test it. It will take some time, return to you tmw.

**Comment by cirdes** [comment]
@marcosmarxm , I added the credentials(secrets.yml) for my meilisearch instance as AIRBYTE_SECRET on github secrets section. Feel free to use if save you time!

**Comment by marcosmarxm** [comment]
> @marcosmarxm , I added the credentials(secrets.yml) for my meilisearch instance as AIRBYTE_SECRET on github secrets section. Feel free to use if save you time!

Can you send them to me in Slack. It is not possible to see a secret again in Github Secrets.

**Comment by cirdes** [comment]
@marcosmarxm, sure!

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**